### PR TITLE
Add integration tests for plugin

### PR DESCRIPTION
**Pull request by yigit**
This PR adds a testing infra for the gradle plugin and also
adds support for android variants.

### CODE REVIEW

**Comment by yigit** [comment]
This is rather a big PR so happy to divide it into more manageable pieces.
I also didn't touch the integration-tests project. It is a bit duplicate but might still makes sense to keep it and use it for full fledged projects while using this one for hand-made cases.

**Comment by yigit** [comment]
re: android integration
I added support for different android specific source sets BUT i am not yet registering the generated folder for the IDE. This PR was getting already too big and has been lurking for a long time so decided to postpone that part for another PR.

**Comment by yigit** [comment]
ping @gavra0 for review. I think you are the only person who has an understanding of how this should work.

File: gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/testing/KspIntegrationTestRule.kt [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+     * like KSP version, local maven repo etc.
+     */
+    val testConfig = TestConfig.read()
+
+    /**
+     * Returns a gradle runner that is ready to run tasks on the test project.
+     */
+    fun runner(): GradleRunner {
+        testProject.writeFiles()
+        return GradleRunner.create()
```
Comment (gavra0) [inline gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/testing/KspIntegrationTestRule.kt:62]: To make sure tests can use `maxParallelForks` property i.e run in parallel, I'd specify GRADLE_USER home env var. Each for gets an id from `System.getProperty("org.gradle.test.worker")` and each for can get the gradle user home location as "$GRADLE_HOME_FOR_TEST/$workerId".

Comment (yigit) [reply gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/testing/KspIntegrationTestRule.kt:62]: i'm running these tests in process (for debuggability) and it is not compatible w/ environment variable passing.

File: gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt [hunk 2]
```diff
@@ -73,13 +82,66 @@ class KspGradleSubplugin @Inject internal constructor(private val registry: Tool
...
+
+    private fun KotlinSourceSet.kspConfiguration(project: Project): Configuration? {
+        return project.configurations.findByName(kspConfigurationName)
+    }
+
     override fun apply(project: Project) {
         project.extensions.create("ksp", KspExtension::class.java)
-        project.configurations.create(KSP_CONFIGURATION_NAME)
-
+        project.pluginManager.withPlugin("org.jetbrains.kotlin.jvm") {
```
[the remark this answers is not in the document]
Comment (yigit) [reply gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt]: done

File: gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/AndroidPluginIntegration.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ * without the Android plugin. The downside is that we need to ensure never to access Android
+ * plugin APIs directly without checking its existence (we have tests covering that case).
+ */
+@Suppress("UnstableApiUsage") // some android APIs are unsable.
+class AndroidPluginIntegration(
+    private val kspGradleSubplugin: KspGradleSubplugin
+) {
+
+    fun applyIfAndroidProject(project: Project) {
+        project.pluginManager.withPlugin("com.android.application") {
```
[the remark this answers is not in the document]
Comment (yigit) [reply gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/AndroidPluginIntegration.kt]: done.

File: gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/testing/KspIntegrationTestRule.kt (hunk 1 above)
Comment (gavra0) [reply gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/testing/KspIntegrationTestRule.kt:62]: I guess that means you are unable to run them in parallel, but that may be ok for now.

**Review by gavra0 — APPROVED** [review]

**Pull request merged**